nfp: Amiibo scanning support

AcK77 opened the pull request:
<div align="center">
<img src="https://user-images.githubusercontent.com/4905390/107304854-2ab91f00-6a82-11eb-9747-f16adb7fce7b.png" />
<br /><i>Thanks to Kennisoan for the logo.</i><br /><br />
<a href="https://w.ryujinx.org/dl/2006"><img src="https://img.shields.io/badge/Download-%232006-black?logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAARCAMAAADjcdz2AAAAe1BMVEVHcEyai5T/XFEAyOr/XlQBxeUBxeX9X1YDxOT/X1UBxeX7XlQAyev7X1QCxeX4YVj+X1UBxeRtmqhnnKv+X1UBxOQCxeUBxOT/X1UAzvELwd8Lwd//VUj+X1U2r8fYbmoAx+j/V0r/X1UCxeb/YlcCzO3/Z1wC1vkC0fPqHM+tAAAAInRSTlMACoFJz879AwP89FBlFeA7o5IZLN+tvZPdFRw0WrZsaX1CnD5B2QAAAI1JREFUGNNNz1cOwyAQBNABO2DAVek9ARvn/icMxSD282k0mkUzyWXXwB93B0wrSxAQ0mSgIx0AwxLwi7WzAssJXmsxHxIQQng9C6FhIlT7LxzoBNXnTTsPYoN7+xhITISOW/uiihfgKhQ2CDtwJV0Egd7IvCNA+1vWPoPbgep4OpeJJn/qwY5ACfrZ/QHYMQ4WY6nIhQAAAABJRU5ErkJggg==&style=flat&labelColor=252a2f&color=f2f5f8" /></a><br /><br />
</div>

All aboard the Amiibo train! 
As promised, this PR fulfills our commitment to provide Amiibo support once the Patreon goal was reached a month ago.

# How does it work?
We currently use a self hosted API (fork of https://github.com/N3evin/AmiiboAPI - big thanks to N3evin!) so you don't have to provide any Amiibo dumps. The benefit of using this API is being able to use all available Amiibo and associated information & images.
Once the game asks for an Amiibo scan, you can go to menu `Options` > `Scan an Amiibo`. A window will pop-up which lets you select the Amiibo you want (thanks to Xpl0itR for the orignal window code).

![Amiibo1](https://user-images.githubusercontent.com/4905390/107304999-66ec7f80-6a82-11eb-8588-7e37f60a82af.PNG)

As you can see, there is a checkbox to allow scanning the same Amiibo more than once per day. And did we mention that writable Amiibo is supported as well? That's right, not only will you be able to use Amiibo for exclusive content in your favorite games, but you can now train your Amiibo in Super Smash Bros. Ultimate! To do that, I've introduced a Virtual Amiibo JSON file which contains all generated data needed for this function.
This current implementation does not support Amiibo raw dump, as it is redundant & unnecessary. But, support for this may be added in the future.

![Amiibo2](https://user-images.githubusercontent.com/4905390/107305019-78358c00-6a82-11eb-9482-c26b194cf391.PNG)
![Amiibo3](https://user-images.githubusercontent.com/4905390/107305021-78358c00-6a82-11eb-9f5d-169f475eb98e.PNG)
![Amiibo4](https://user-images.githubusercontent.com/4905390/107305022-78ce2280-6a82-11eb-8528-b6254abdd154.PNG)
![Amiibo5](https://user-images.githubusercontent.com/4905390/107305016-779cf580-6a82-11eb-9900-257d83d437c6.PNG)

Please test all Amiibo features in as many games as possible, and let us know how it's working in the comments or on Discord. 

Note to other devs:
- Fixes #1472
- The service part is checked by RE.
~~- There is a little surprise yet to come before this PR is ready to merge.~~

![Amiibo6](https://user-images.githubusercontent.com/4905390/107305062-84214e00-6a82-11eb-8c5f-92a47197bf9b.PNG)
![Amiibo7](https://user-images.githubusercontent.com/4905390/107305065-85527b00-6a82-11eb-9d00-ba9ec9cdeef0.PNG)
![Amiibo8](https://user-images.githubusercontent.com/4905390/107305052-81265d80-6a82-11eb-9e9d-676ac2c675b4.PNG)

jduncanator commented:
Marking as draft until final changes are committed to avoid an accidental merge :)

Thog commented:
I don't think there will be accidental merge here @jduncanator considering 2 reviews are required, undrafting

Shahil-Ayato commented:
![18D8B238-B206-4E2C-9D22-610C60B33002](https://user-images.githubusercontent.com/67567036/107334632-d014c880-6a6b-11eb-9ef8-ea46ba4da55d.png)
I feel that the amiibo should have its own tab in top right or be put in the {Tools} tab

CeruleanSky commented on Ryujinx/Ui/Windows/AmiiboWindow.cs:
```diff
@@ -0,0 +1,276 @@
+        {
+            try
+            {
+                HttpResponseMessage response = await _httpClient.GetAsync("https://www.google.com/");
```
Google might not be available everywhere and connecting to it doesn't mean you can connect to ryujinx.org. 
Perhaps:
`_httpClient.SendAsync(new HttpRequestMessage (HttpMethod.Head, "https://amiibo.ryujinx.org/"));`
To just read headers, thus making a light request on the ryujinx servers. 
Then the error message can say something along the lines of
`"Unable to connect to Amiibo information server, the service is down or check your internet connection"`

Benedani replied:
Checking for a different website when the Ryujinx ping failed could still differentiate between your internet and the service being down. Actually, why not just ping the web server instead of connecting to it?

CeruleanSky replied:
> Actually, why not just ping the web server instead of connecting to it?

Pinging is light and would let you know if your network stack can communicate with the network stack of the machine whose ip address is resolved, but wouldn't tell you if the amiibo.ryujinx.org http server is functioning and connectable , which is going to happen sooner or later if the ping works.

The suggestion checks your internet is working, can connect, the http server is connectable, and with the `last-modified` field can even let you know if you have the latest version all in one step.

>Checking for a different website when the Ryujinx ping failed could still differentiate between your internet and the service being down.

If needed, that would certainly help diagnose things after the failure.  But a proper website would need to be picked and so forth, instead I just suggested changing the error message to recommend the user (manually) check their internet or let them assume the server was down.

GamerzHell9137 commented:
Here's some issues and suggestions i'd like to give:

1. Hyrule Warriors crashes when Amiibo is chosen because of an unimplemented service
```00:01:41.113 |W| HLE.OsThread.5 KernelIpc CallMethod: Missing service Ryujinx.HLE.HOS.Services.Nfc.IUserManager: 0 ignored
```
2. I'd change the drop down list style to a text box style ala System TimeZone since some Amiibo series have way too many Amiibos (Animal Crossing) 
![image](https://user-images.githubusercontent.com/2002038/107347932-a5a42900-6ac6-11eb-86e2-b4ddfa3c36a5.png)
![image](https://user-images.githubusercontent.com/2002038/107347973-b359ae80-6ac6-11eb-9388-b23cf378e828.png)

3. If not changed then the UI should be fixed, things above in the AC list is empty and it fills up the whole screen, vertically both above and under legit entries. example of empty lines above
![image](https://user-images.githubusercontent.com/2002038/107348116-e603a700-6ac6-11eb-9ee5-bc7b7bc34f00.png)

4. A way to press a letter on keyboard so that the list drops down to that section of the list (again, for games that have way too many amiibos ;Animal Crossing)

6Fer6 commented:
Works flawlessly on the following games: Super Mario Odyssey, Splatoon 2, Super Smash Bros Ultimate, Captain Toad: Treasure Tracker, Yoshi’s Crafted World and Mario Kart 8 Deluxe. But there are some issues with the GUI:
•	In Super Mario Odyssey, if you press the right dpad button it allows you to use amiibo, then press the Options button in Ryujinx it shows the ‘Scan an Amiibo’ in white, but if you release the right dpad button, the ‘Scan an Amiibo’ option is still clickable. (This is a limitation with GTK, but I was told to comment on the PR)
![image](https://user-images.githubusercontent.com/70550490/107408290-8965b380-6ae9-11eb-8fa2-74c94fcf4d23.png)
•	Currently every time you open the ‘Scan an Amiibo’ option it defaults to 8-Bit Mario. It would be a nice if you previously scanned a character, it would automatically appear the same character on the list again. This would be useful for games like Splatoon 2, where you must scan a character for a second time to add them as friends, and SSBU, where you must scan again to save all your training.
•	Move the button to the bar on the top, for easier access.  
![image](https://user-images.githubusercontent.com/70550490/107408342-9a162980-6ae9-11eb-9135-96ac60004e51.png)

Jerr-z commented:
super smash bros does not recognize link amiibo, and attempting to edit the name for amiibo fighter will result in a crash

EmulationFanatic commented:
> super smash bros does not recognize link amiibo, and attempting to edit the name for amiibo fighter will result in a crash

Confirmed both of these things, though the Link Amiibo issue is the only surprising one. Lack of custom user profile support is the reason for the name editing crash.

Jerr-z commented:
another issue observed in smash ultimate is if you scan the amiibo from the amiibo menu and choose battle the amiibo, you cannot save the amiibo data after the battle because the game suggests that the data is overwritten

komawoyo commented:
So if I'm out in the boonies without internet, this will not work?

Thog reviewed (approved):
lgtm

PokemongamerEB commented:
Every time I try to use an amiibo in Smash Ultimate, the game crashes. Any idea why?

PokemongamerEB commented:
> Every time I try to use an amiibo in Smash Ultimate, the game crashes. Any idea why?

Also if I need to send any info I will

AcK77 commented:
Come to our discord and contact on it, it will be easier than GitHub comments because we will need more informations for sure. Thanks.

PokemongamerEB commented:
> Come to our discord and contact on it, it will be easier than GitHub comments because we will need more informations for sure. Thanks.

What's the discord?

CeruleanSky commented:
> • In Super Mario Odyssey, if you press the right dpad button it allows you to use amiibo, then press the Options button in Ryujinx it shows the ‘Scan an Amiibo’ in white, but if you release the right dpad button, the ‘Scan an Amiibo’ option is still clickable. (This is a limitation with GTK, but I was told to comment on the PR)

To add to the above comment about Mario Odyssey, at least for me in win 10, if you are using keyboard for input, you can't click on the menu at all.  I held the right arrow key to open the amiibo scan prompt inside the game, and I could no longer use the Ryujinx toolbar.

Mou-Ikkai commented:
Can .bin support be added? Can be useful for when I want to use my previously trained SSBU amiibos on Ryujinx.

jduncanator commented:
> Can .bin support be added? Can be useful for when I want to use my previously trained SSBU amiibos on Ryujinx.

It was outlined in the PR description that this isn't a priority or a target for this specific PR, however anyone is free to add that functionality in the future.

CeruleanSky commented:
> Can .bin support be added? Can be useful for when I want to use my previously trained SSBU amiibos on Ryujinx.

From the opening post "This current implementation does not support Amiibo raw dump, as it is redundant & unnecessary. But, support for this may be added in the future." , 

I suppose the discord would have the latest info, gbatemp has some forks of emuiibo that can convert them, though there might be better utilities now.

AcK77 commented:
- Added a new Actions menu, under which the functions Stop Emulation, Simulate Wake-up Message and Scan an Amiibo have been moved.
![unknown1](https://user-images.githubusercontent.com/4905390/111011795-d82d9400-839a-11eb-8798-c426fe5bc973.png)

- Removed usage of Google to check connectivity. Instead it now uses the "last-modified" header instead (suggested by @CeruleanSky).
- Only Amiibo related to the currently running game are shown, along with a pertinent usage description and indicator of whether the Amiibo is writable.
  - Added a checkbox option to show all Amiibo if the user wishes.
  ![unknown2](https://user-images.githubusercontent.com/4905390/111011810-e8de0a00-839a-11eb-8a5c-d03b0e094ec4.png)

- Added a message dialog when the Amiibo scanning isn't available in guest game (Fixes SMO DPad issue, reported by @6Fer6).
- The last scanned Amiibo is now stored at each game execution, which eases use of scanning writable Amiibo (suggested by @6Fer6).
- Fixed Amiibo ID format: now every Amiibo should be recognized properly (Fixes SSBU issue, reported by @Jerr-z).
  ![unknown3](https://user-images.githubusercontent.com/4905390/111011840-01e6bb00-839b-11eb-9160-f86f4f5a81e7.png)


@GamerzHell9137 Those extra spaces in the dropdown menus are due to a GTK issue and I'll not change it to something else. The list is already reduced now.

Hyrule Warriors crashes when Amiibo is chosen because of an unimplemented service isn't fixed for now, it will maybe fixed once this PR is merged.
```
00:01:41.113 |W| HLE.OsThread.5 KernelIpc CallMethod: Missing service Ryujinx.HLE.HOS.Services.Nfc.IUserManager: 0 ignored
```

@Jerr-z I'm not able to reproduce this issue:
```
another issue observed in smash ultimate is if you scan the amiibo from the amiibo menu and choose battle the amiibo, you cannot save the amiibo data after the battle because the game suggests that the data is overwritten
```

I'd want to add Emuiibo files support (as a surprise), Ryujinx currently got a license exemption from XorTroll (https://github.com/XorTroll/emuiibo/commit/a442cc117bf1279e56446c15884d656e91a71c59), Thanks to him.
But this take time to implement and I don't want to block this PR anymore so it will be implemented later.

stromcon commented:
Hey there,

tried on MK8 working fine !

![Capture_decran_2021-03-13_132637](https://user-images.githubusercontent.com/6690149/111030800-7c97f080-8404-11eb-95c3-82a16af22aee.png)
![Capture_decran_2021-03-13_132556](https://user-images.githubusercontent.com/6690149/111030801-7d308700-8404-11eb-8df7-113b944bb860.png)

But on FE, when pressing "scan it!" button nothing happens, even with hack checkbox. game is not crashing, we can go back by pressing B and console logs are not reporting anything. Here is the video with the test on FE (using discord link because MP4 is > 10 Mb) : https://discord.com/channels/410208534861447168/814263228926132274/820279075036725288

Thanks for your work !

makaragamz commented:
Sorry but I may be missing something. I tried both Downloading the latest release from the website and also build the master branch (dirty) but I still have no option for Loading Amiibos in the Options menu, do I have to do any additional step?

EmulationFanatic commented:
@makaragamz it's in the new Actions menu as shown here https://github.com/Ryujinx/Ryujinx/pull/2006#issuecomment-797829812

makaragamz commented:
> @makaragamz it's in the new Actions menu as shown here [#2006 (comment)](https://github.com/Ryujinx/Ryujinx/pull/2006#issuecomment-797829812)

Oh I'm terribly sorry, didn't notice that comment, I just tried from there and its working fine, thank you very much! I will keep testing this feature, where can I share my feedback on its functionality? I'd like to help with my findings

HarshNarayanJha commented:
It says Unable to connect to API server, and I have Internet connection and even tried with firewall disabled.

M3morize0104 commented:
Sorry, it's not clear to me how I'm supposed to install this. I'm not a programmer or anything like that, could you explain in a more general and simple way the steps?

AcK77 commented:
https://github.com/Ryujinx/Ryujinx/wiki/Ryujinx-Setup-&-Configuration-Guide#amiibo-emulation
